DatadogCrashReporting/Sources/KSCrashIntegration/DatadogDiagnosticFilter.swift

```diff
@@ -0,0 +1,170 @@
+    ///
+    /// This method analyzes each crash report, generates a diagnostic message describing
+    /// the crash cause, and injects it into the report. The diagnosis is added to both
+    /// the main crash report and any recrash reports present.
+    ///
+    /// - Parameters:
+    ///   - reports: The crash reports to process and enhance with diagnostic information
+    ///   - onCompletion: Completion handler called with the processed reports. If diagnosis
+    ///                   generation fails for any report, the original reports are returned
+    ///                   along with the error.
+    func filterReports(
+        _ reports: [CrashReport],
+        onCompletion: (([CrashReport]?, (Error)?) -> Void)?
+    ) {
+        do {
+            let reports = try reports.map { report in
+                // Validate and extract the type-safe report dictionary
+                guard var dict = report.untypedValue as? CrashFieldDictionary else {
+                    throw CrashReportException(description: "KSCrash report untypedValue is not a CrashDictionary")
+                }
+
+                if let crash: CrashFieldDictionary = try dict.valueIfPresent(forKey: .crash), let diagnosis = try diagnose(crash: crash) {
+                    dict.setValue(forKey: .crash, .diagnosis, value: diagnosis)
+                }
+
+                if let crash: CrashFieldDictionary = try dict.valueIfPresent(forKey: .recrashReport, .crash), let diagnosis = try diagnose(crash: crash) {
+                    dict.setValue(forKey: .recrashReport, .crash, .diagnosis, value: diagnosis)
+                }
+
+                return AnyCrashReport(dict)
+            }
+
+            onCompletion?(reports, nil)
+        } catch {
+            onCompletion?(reports, error)
+        }
+    }
+
+    /// Analyzes a crash report and generates a human-readable diagnostic message.
+    ///
+    /// This method examines the crash data to determine the crash type and constructs
+    /// an appropriate diagnostic message. It follows a prioritized analysis:
+    ///
+    /// 1. **Exception-based crashes**: Checks for uncaught NSException
+    /// 2. **Signal-based crashes**: Checks for Unix signals with known descriptions
+    /// 3. **Unknown crashes**: Returns a generic message for unrecognized crash types
+    ///
+    /// - Parameter dict: The crash report dictionary containing crash information
+    /// - Returns: A diagnostic message describing the crash, or `nil` if no crash data is found
+    /// - Throws: `CrashReportException` if the report structure is invalid
+    ///
+    /// ## Example Messages
+    ///
+    /// - Exception: `"Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'unrecognized selector'."`
+    /// - Signal: `"Application crash: SIGSEGV (Segmentation fault)"`
+    /// - Unknown: `"Application crash: <unknown>"`
+    func diagnose(crash dict: CrashFieldDictionary) throws -> String? {
```

Comment:
nit: it seems that this method is always returning a string 
```suggestion
    func diagnose(crash dict: CrashFieldDictionary) throws -> String {
```